src/ripple/app/tx/impl/SetOracle.cpp

```diff
@@ -0,0 +1,259 @@
+    {
+        // update
+
+        // lastUpdateTime must be more recent than the previous one
+        if (ctx.tx[sfLastUpdateTime] <= (*sle)[sfLastUpdateTime])
+            return tecINVALID_UPDATE_TIME;
+
+        if (ctx.tx[sfAccount] != sle->getAccountID(sfOwner))
+            return tecNO_PERMISSION;
+
+        if (ctx.tx.isFieldPresent(sfProvider) ||
+            ctx.tx.isFieldPresent(sfAssetClass))
+            return temMALFORMED;
+
+        for (auto const& entry : sle->getFieldArray(sfPriceDataSeries))
+        {
+            auto const hash = tokenPairHash(entry);
+            if (!pairs.contains(hash))
+                pairs.emplace(hash);
+        }
+    }
+    else
+    {
+        // create
+
+        if (!ctx.tx.isFieldPresent(sfProvider) ||
+            !ctx.tx.isFieldPresent(sfAssetClass))
+            return temMALFORMED;
+    }
+
+    if (pairs.size() > maxOracleDataSeries)
+        return temBAD_ARRAY_SIZE;
+
+    auto const add = pairs.size() > 5 ? 2 : 1;
+    auto const reserve = ctx.view.fees().accountReserve(
+        sleSetter->getFieldU32(sfOwnerCount) + add);
+    auto const& balance = sleSetter->getFieldAmount(sfBalance);
+
+    if (balance < reserve)
+        return tecINSUFFICIENT_RESERVE;
+
+    return tesSUCCESS;
+}
+
+static bool
+adjustOwnerCount(ApplyContext& ctx, std::uint16_t count)
+{
+    if (auto const sleAccount =
+            ctx.view().peek(keylet::account(ctx.tx[sfAccount])))
+    {
+        adjustOwnerCount(ctx.view(), sleAccount, count, ctx.journal);
+        return true;
+    }
+
+    return false;
+}
+
+TER
+SetOracle::doApply()
+{
+    auto const oracleID = keylet::oracle(account_, ctx_.tx[sfOracleDocumentID]);
+
+    if (auto sle = ctx_.view().peek(oracleID))
+    {
+        // update
+        // the token pair that doesn't have their price updated will not
+        // include neither price nor scale in the updated PriceDataSeries
+
+        hash_map<uint256, STObject> pairs;
+        // collect current token pairs
+        for (auto const& entry : sle->getFieldArray(sfPriceDataSeries))
+        {
+            STObject priceData{sfPriceData};
+            priceData.setFieldCurrency(
+                sfBaseAsset, entry.getFieldCurrency(sfBaseAsset));
+            priceData.setFieldCurrency(
+                sfQuoteAsset, entry.getFieldCurrency(sfQuoteAsset));
+            pairs.emplace(tokenPairHash(entry), std::move(priceData));
+        }
+        auto const oldCount = pairs.size() > 5 ? 2 : 1;
+        // update/add pairs
+        for (auto const& entry : ctx_.tx.getFieldArray(sfPriceDataSeries))
+        {
+            auto const hash = tokenPairHash(entry);
+            if (auto iter = pairs.find(hash); iter != pairs.end())
+            {
+                // update the price
+                iter->second.setFieldU64(
+                    sfAssetPrice, entry.getFieldU64(sfAssetPrice));
+                if (entry.isFieldPresent(sfScale))
```

Comment:
I'm confused about the usage of `Scale` factor. The spec says that `sfAssetPrice` is the "scaled asset price, which is the price value after applying the scaling factor".

If I already have the final scaled price for an asset, why would I need the `scale` information? I do not need the `scale` information to determine the cost of say, 10 units of the base asset. Am I missing something?

![image](https://github.com/XRPLF/rippled/assets/21219765/293e3d68-2c55-4db1-8e38-112062c0eb57)

Reply:
Suppose the `AssetPrice` is 10,000 and the `QuoteAsset` is `USD` then is the price $10,000 or $1000 or $100, ...? The `Scale` tells you what the actual price is. If the `Scale` is 2, then the price is $100.

Reply:
Yes.

Reply:
`SymbolPrice` without the `Scale` is not complete. Let's say I'm a dApp and I'm looking at the pair `BTC/USD` with the `SymbolPrice` 43,379,739,841`USD`. What does 43,379,739,841 mean? The `Scale` of 6 makes it clear that one BTC is 43,379.739841`USD`. The `originalPrice` is essentially represented in the Price Oracle as mantissa and exponent with the exponent implied to be always negative and the exponent only used to represent the fractional. So if one `BTC` is 43,379`USD` then the `SymbolPrice` is 43,379 and the `Scale` is 0. The Oracle provider might choose to always represent `BTC` with the `Scale` 6, in which case one `BTC` is represented with the `SymbolPrice` 43,379,000,000`USD` and the `Scale` 6.

Reply:
okay, this makes a lot of sense. thank you for the explanation!!